pkg/sql/catalog/internal/catkv/catalog_reader.go

```diff
@@ -228,18 +228,62 @@ func (cr catalogReader) GetByIDs(
 		isDescriptorRequired: isDescriptorRequired,
 		expectedType:         expectedType,
 	}
-	err := cq.query(ctx, txn, &mc, func(codec keys.SQLCodec, b *kv.Batch) {
-		for _, id := range ids {
+	// Tracks the start and end of the run of descriptor ID's.
+	startIDSet := false
+	runStartID := descpb.InvalidID
+	runEndID := descpb.InvalidID
+	// Emit batch will generate either individual get calls or
+	// scans for batches of descriptors. In certain benchmarks
+	// with a large number of descriptors accessed sequentially,
+	// this can be beneficial for the crdb_internal functions.
+	emitBatch := func(codec keys.SQLCodec, b *kv.Batch) {
+		// Only a single descriptor run, so generate a get requst.
+		if runStartID == runEndID {
+			id := runStartID
 			get(ctx, b, catalogkeys.MakeDescMetadataKey(codec, id))
 			for _, t := range catalogkeys.AllCommentTypes {
 				scan(ctx, b, catalogkeys.MakeObjectCommentsMetadataPrefix(codec, t, id))
 			}
 			get(ctx, b, config.MakeZoneKey(codec, id))
+		} else {
+			// Otherwise, generate a scan request instead. The end key is exclusive,
+			// so we will increment the end key by one.
+			scanRange(ctx, b, catalogkeys.MakeDescMetadataKey(codec, runStartID),
+				catalogkeys.MakeDescMetadataKey(codec, runEndID+1))
+			for _, t := range catalogkeys.AllCommentTypes {
+				scanRange(ctx, b, catalogkeys.MakeObjectCommentsMetadataPrefix(codec, t, runStartID),
+					catalogkeys.MakeObjectCommentsMetadataPrefix(codec, t, runEndID+1))
+			}
+			scanRange(ctx, b, config.MakeZoneKey(codec, runStartID),
+				config.MakeZoneKey(codec, runEndID+1))
+		}
+		startIDSet = false
+	}
+
+	err := cq.query(ctx, txn, &mc, func(codec keys.SQLCodec, b *kv.Batch) {
+		for _, id := range ids {
```

Comment:
I think this works. But I also think it would be nice to have a function to calculate the continuous segments instead of having the function scope variables which can be hard to interpret in the future :) Nice work though!

Reply:
Done, pulled the logic into a function